### ISSUE TITLE
[IDEA] Introduce a way to group items in Control Panel tabs

### ISSUE THREAD

**Issue by saqimtiaz** [Jermolene/TiddlyWiki5#4957]
Background: #4952 

We need a better way to organize and group keyboard shortcuts, settings etc in the control panel tab. 

For instance the keyboard shortcuts list is long and hard to scan at a glance.
Perhaps a way to group them: collapsible sections or sub-tabs for editor shortcuts, UI shortcuts, etc.
`toggle-sidebar` for example, is listed between `superscript` and `underline`, which isn't intuitive when you don't know how the keyboard shortcut might be described.

@Jermolene suggested that: "The palette editor and settings tab suffer from many of the same problems, perhaps there's scope to bring in a general purpose component that groups items and provides dynamic search (similar to the existing plugins listing)."

**Comment by rmunn** [comment]
In addition to grouping the shortcuts by category or context (editor, story view, or whatever), I'd like to also see two other groups. One would be a group called "All shortcuts by name" that presents them in alphabetical order of the shortcut name, i.e. exactly the way they're currently presented. And the other would be a group called "All shortcuts by key", that presents them in pseudo-alphabetical order of the keys used, with the following guidelines:

* Keys with modifiers are grouped together with the same key without modifiers, so that Enter, ctrl-Enter, and shift-Enter are all next to each other in the list. Within one group (say, the Enter group), the "base" key (Enter) always sorts first, then keys with a single modifier sort next in "standard" order of modifiers (defined below), then all two-modifier keys in "standard" order of the first modifier followed by "standard" order of the second, and so on for any shortcuts with three or more modifiers.
    * "Standard" modifier order is ctrl comes first, then shift, then alt comes last. Just like how you write ctrl-shift-A, not shift-ctrl A, you write shift-alt-S, not alt-shift-S, and you write ctrl-alt-Del, not alt-ctrl-Del.
    * Example because the last sentence of that bullet point is rather dense: if all the following shortcuts existed, they would sort in the following order: Enter, ctrl-Enter, shift-Enter, alt-Enter, ctrl-shift-Enter, ctrl-alt-Enter, shift-alt-Enter, ctrl-shift-alt-Enter.
* Keys that are not single letters (Enter, Escape, etc) sort alphabetically in a *different section* than the letter keys. E.g., if there are ctrl-E, ctrl-F, ctrl-G, F2, and Enter shortcuts in someone's wiki, I would not want to see them sort as ctrl-E, then Enter, then ctrl-F, then F2, then ctrl-G. I don't know yet if keys that are not letters should sort before or after letter keys, but I am leaning towards "before", in which case the right sort order of those five keys would be Enter, F2, ctrl-E, ctrl-F, ctrl-G.

Open questions about the design I just outlined:
* What about the Mac modifier key Cmd (⌘)? Where should it sort in the ctrl-shift-alt order? I don't use a Mac, so I have no experience with how keyboard shortcuts are typically presented in OS X when ⌘ is combined with another modifier. Do they write it as ⌘-Shift-A, or Shift-⌘-A, or Shift-⌘A, or what? If someone who uses a Mac could look in the menu of their various apps and determine what the sorting rule is for ⌘ when combined with other modifiers, that would be useful to help us determine how to present that key.
* What about the Mac "option" key (⌥)? Is it even possible for applications (and webpages running in the browser) to see the ⌥ key and use it for anything? Or is it exclusively captured by OS X and used to allow typing useful Unicode characters, such as option-v to get the √ symbol, and apps can't override that? If OS X captures the option key and doesn't let apps change its behavior, then we can never have opt-V as a keyboard shortcut and we don't need to worry about how it sorts. But if apps can assign behavior to combinations like opt-V, then we also need to find out how it sorts. Again, if someone who uses a Mac could look in the menu of their various apps and determine what the sorting rule is for the option key (⌥) when combined with other modifiers, that would be useful to help us determine how to present that key.

**Comment by saqimtiaz** [comment]
@rmunn all good suggestions, thank you.

**Comment by kookma** [comment]
###  [IDEA] Overhaul of Tiddlywiki Settings #5124

**Is your feature request related to a problem? Please describe.**
Right now all the settings can be accessed by opening the $:/ControlPanle. Are all of these required for normal users?  Of course there is an Advanced Tab which used by superusers.

**Describe the solution you'd like**
A propose a setting page like Chrome

![image](https://user-images.githubusercontent.com/830394/100337247-63ea0300-2fec-11eb-9f19-719cce8d9af4.png)

It has three features:

1. A searchbox to find a setting (or hidden one)
2. A Tree like menu in the left with icons to simply recognize categories. Click on each category open the setting for that category in the right panel.
3. A hidden Advanced section on click expand advanced setting categories

I think replicating the Chrome Setting page in Tiddlywiki is simple. TW has all required tools to create such elements.

If we want to keep the current design, I am insisted to have item 1 and 2 above.

**Comment by kookma** [comment]
Some other Setting pages to get idea

#### Windows 10
![109_ApplicationFrameHost](https://user-images.githubusercontent.com/830394/100339423-19b65100-2fef-11eb-85de-f80a5868aeb8.png)


Note the  search in setting on the top right

#### GitHub
![108_%pn](https://user-images.githubusercontent.com/830394/100339464-263aa980-2fef-11eb-9f1a-bdd54888e350.png)

**Comment by kookma** [comment]
Note: Using sub pages like figures above makes Tiddlywiki Control Panel pages work better on small screen devices.

Worth have a look on the below discussions

- https://ux.stackexchange.com/questions/48214/best-layout-for-a-settings-page
- https://uxdesign.cc/designing-a-better-settings-page-for-your-app-fcc32fe8c724
- http://designinginterfaces.com/patterns/settings-editor/

**Comment by saqimtiaz** [comment]
Also see #4507 wit ideas about search and find in settings.

**Comment by twMat** [comment]
This might be a good time and context to consider "searchable keywords", i.e words that are specifically added for the sake of findability but that are not (necessarily) part of the tiddlers text or the tags.

I've called for a "keywords" feature previously and created [searchwords](http://searchwords.tiddlyspot.com/). The Controlpanel is a *particularly* good use case because all users must deal with the controlpanel but at the same time it is very "lingo heavy". For example; a user who naively searches for either of "color" or "colour" should, of course, get the *palette* as one of the search results and probably also the *themes*. And users should ideally also be able to add custom keywords to controlpanel items for findability, e.g "Fav. settings", "Disability settings", etc and find all the concerned settings in one search.

The Controlpanel equivalents in Win 10, Chrome browser and more, allow this type of free text search to locate settings.

Hope this was not off topic.

**Comment by saqimtiaz** [comment]
@twMat yes this thought has occurred to me as well when thinking of adding the ability to search within the Control Panel. I am planning to work on a settings UI for my Streams plugin after the TW 5.1.23 release and use that as an opportunity to test some UI ideas for the core control panel.
